baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/ExceptionSpecificity.java

```diff
@@ -0,0 +1,411 @@
+                ImmutableList<Type> thrownCheckedExceptions = normalizeExceptions(
+                        getThrownCheckedExceptions(tree, state), state);
+                if (containsBroadException(thrownCheckedExceptions, state)) {
+                    return Description.NO_MATCH;
+                }
+                ImmutableList<Type> thrown = flattenExceptionTypes(thrownCheckedExceptions, state);
+                if (thrown.size() <= MAX_CHECKED_EXCEPTIONS
+                        // Do not apply this to test code where it's likely to be noisy.
+                        // In the future we may want to revisit this.
+                        && !TestCheckUtils.isTestCode(state)) {
+                    List<Type> replacements = deduplicateCatchTypes(
+                            ImmutableList.<Type>builder()
+                                    .addAll(thrown)
+                                    .addAll((isThrowable ? THROWABLE_REPLACEMENTS : EXCEPTION_REPLACEMENTS).stream()
+                                            .map(name -> Preconditions.checkNotNull(
+                                                    state.getTypeFromString(name), "Failed to find type"))
+                                            .collect(ImmutableList.toImmutableList()))
+                                    .build(),
+                            encounteredTypes,
+                            state);
+                    SuggestedFix.Builder fix = SuggestedFix.builder();
+                    if (replacements.isEmpty()) {
+                        // If the replacements list is empty, this catch block isn't reachable and can be removed.
+                        fix.replace(catchTree, "");
+                    } else {
+                        Name parameterName = catchTree.getParameter().getName();
+                        AssignmentScanner assignmentScanner = new AssignmentScanner(parameterName);
+                        catchTree.getBlock().accept(assignmentScanner, null);
+                        if (replacements.size() == 1 || !assignmentScanner.variableWasAssigned) {
+                            catchTree.accept(new ImpossibleConditionScanner(
+                                    fix, replacements, parameterName), state);
+                            fix.replace(catchTypeTree, replacements.stream()
+                                    .map(type -> SuggestedFixes.prettyType(state, fix, type))
+                                    .collect(Collectors.joining(" | ")));
+                        }
+                    }
+                    return buildDescription(catchTypeTree)
+                            .addFix(fix.build())
+                            .build();
+                }
+                return Description.NO_MATCH;
+            }
+            // mark the type as caught before continuing
+            encounteredTypes.add(catchType);
+        }
+        return Description.NO_MATCH;
+    }
```

Comment:
I think i'm still struggling with this method - there's quite a lot of intricacy in the de-duping and selecting only exceptions&throwables that is all woven together to get this all done in one pass.

what about if we did loop through all the possible `tree.getCatches()` to accumulate the `encounteredTypes`, then could call out to a dedicated function to process each single catch?

Reply:
also the possibility of fully returning from deep inside a loop & if condition feels a bit weird - are we saying if we hit one `catch(Exception)` then a later `catch(Throwable)` might not get processed?

Reply:
> are we saying if we hit one catch(Exception) then a later catch(Throwable) might not get processed?

Yep, we have a test for that case which takes two rounds to resolve. We can do it in one loop, but I worried about the additional complexity. I can take a look at implementing it and we can decide from there.

Reply:
Personally I'm fine with it just doing one thing at a time - minimum complexity = happy dan. Just wanted to check my understanding.

Reply:
Turns out it's straightforward to fix 'em all at once. I think that blocks us from splitting into multiple loops though because the changes mutate state, and we would have to keep expected handled types in sync with the code that builds suggested fixes, leading us back to the original model.